### PR TITLE
[CostModel][X86] Add initial costs for non-lane-crossing one/two input shuffles

### PR DESCRIPTION
**Pull request by RKSimon**
Most of the x86 shuffle instructions operate within each 128-bit subvector lane, but our shuffle costs struggle to handle this and have to fallback to worst case shuffles that reference elements from any lane.

This patch detects shuffle masks that we know are "inlane" and enable us to assume a cheaper shuffle cost.

### CODE REVIEW

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 1]
```diff
@@ -1559,6 +1559,23 @@ InstructionCost X86TTIImpl::getShuffleCost(
...
+      (BaseTp->getPrimitiveSizeInBits() % 128) == 0 &&
+      Mask.size() == BaseTp->getElementCount().getKnownMinValue()) {
+    unsigned NumLanes = BaseTp->getPrimitiveSizeInBits() / 128;
+    unsigned NumEltsPerLane = Mask.size() / NumLanes;
+    if ((Mask.size() % NumLanes) == 0) {
+      IsInLaneShuffle = true;
+      for (auto [I, M] : enumerate(Mask))
+        if (0 <= M)
+          IsInLaneShuffle &=
+              ((M % Mask.size()) / NumEltsPerLane) == (I / NumEltsPerLane);
```
Comment (alexey-bataev) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1575]: `IsInLaneShuffle = all_of(enumerate(Mask), [&](const auto &P) { return P.value() == PoisonMaskElem || ((P.value() % Mask.size()) / NumEltsPerLane) == (P.index() / NumEltsPerLane); });`

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 2]
```diff
@@ -1559,6 +1559,22 @@ InstructionCost X86TTIImpl::getShuffleCost(
       return TTI::TCC_Free;
   }
 
+  // Attempt to detect a cheaper inlane shuffle, avoiding 128-bit subvector
+  // permutation.
+  bool IsInLaneShuffle = false;
+  if (BaseTp->getPrimitiveSizeInBits() > 0 &&
+      (BaseTp->getPrimitiveSizeInBits() % 128) == 0 &&
+      Mask.size() == BaseTp->getElementCount().getKnownMinValue()) {
```
Comment (alexey-bataev) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1567]: Shall we also check that the original type of the vector element is the same as the type of the "legalized" (machine) vector type?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1567]: Sure, I think checking the scalar sizes match should be enough - I'll add that shortly

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 3]
```diff
@@ -1564,6 +1564,7 @@ InstructionCost X86TTIImpl::getShuffleCost(
   bool IsInLaneShuffle = false;
   if (BaseTp->getPrimitiveSizeInBits() > 0 &&
       (BaseTp->getPrimitiveSizeInBits() % 128) == 0 &&
+      BaseTp->getScalarSizeInBits() == LT.second.getScalarSizeInBits() &&
```
Comment (alexey-bataev) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1567]: Will it work for pointers?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp:1567]: Not, thats what `BaseTp->getPrimitiveSizeInBits() > 0` is checking for as ptr vectors return 0

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**